### PR TITLE
tests: add PID verifications with commit/rollback

### PR DESCRIPTION
**Pull request by blankoworld**
Record that failed validation generates a PID. It should not.

If any validation error occurs during a Record validation,
the PID should be canceled in database (using a rollback).

Some tests have been done to check this.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/api.py [hunk 1]
```diff
@@ -137,7 +138,12 @@ def create(cls, data, id_=None, delete_pid=False,
         if not id_:
             id_ = uuid4()
         cls.minter(id_, data)
-        record = super(IlsRecord, cls).create(data=data, id_=id_, **kwargs)
+        # We don't need to keep the PID if any validation error
+        try:
+            record = super(IlsRecord, cls).create(data=data, id_=id_, **kwargs)
+        except (RecordValidationError, ValidationError) as e:
+            db.session.rollback()
```
Comment (rerowep) [inline rero_ils/modules/api.py]: I'am not sure this is atomic. I think we have to use `with db.session.begin_nested():`

File: rero_ils/modules/api.py [hunk 2]
```diff
@@ -137,7 +138,12 @@ def create(cls, data, id_=None, delete_pid=False,
         if not id_:
             id_ = uuid4()
         cls.minter(id_, data)
-        record = super(IlsRecord, cls).create(data=data, id_=id_, **kwargs)
+        # We don't need to keep the PID if any validation error
+        try:
+            record = super(IlsRecord, cls).create(data=data, id_=id_, **kwargs)
+        except (RecordValidationError, ValidationError) as e:
```
Comment (rerowep) [inline rero_ils/modules/api.py]: One letter variables `e` are not recommended.

File: rero_ils/modules/api.py [hunk 3]
```diff
@@ -137,7 +138,12 @@ def create(cls, data, id_=None, delete_pid=False,
...
+            db.session.rollback()
+            raise e
```
Comment (rerowep) [inline rero_ils/modules/api.py]: Maybe `raise` will interrupt the import of fixtures. We have to be clear if we like to interrupt the import or if we have to intercept the exception and display a error message for example.

File: rero_ils/modules/api.py (hunk 1 above)
Comment (blankoworld) [reply rero_ils/modules/api.py]: I think so too. But then I check https://github.com/inveniosoftware/invenio-records/blob/master/invenio_records/api.py#L172-L190 which already uses `db.session.begin_nested()`.

That's why I just use `db.session.rollback()`.

File: rero_ils/modules/api.py (hunk 3 above)
Comment (blankoworld) [reply rero_ils/modules/api.py]: Which fixtures do you talk about? Those from `./scripts/setup`?

I will check that.

Comment (blankoworld) [reply rero_ils/modules/api.py]: On my side, it seems to be OK.

File: tests/ui/test_api.py [hunk 4]
```diff
@@ -201,3 +203,64 @@ def test_ilsrecord(app, es_default_index, ils_record, ils_record_2):
...
+FailedPidProvider = type(
+    'FailedPidProvider',
+    (Provider,),
+    dict(
+        identifier=FailedPidIdentifier,
+        pid_type='failed'
+    )
+)
+
+# cipo minter
```
Comment (rerowep) [inline tests/ui/test_api.py]: What is `cipo`

File: rero_ils/modules/api.py (hunk 1 above)
Comment (blankoworld) [reply rero_ils/modules/api.py]: Minter add the pid to the session here: https://github.com/inveniosoftware/invenio-pidstore/blob/master/invenio_pidstore/models.py#L562
The `with` statement should commit at the end. So PID should be reserved.

But I didn't notice that the PID remains after the rollback yet. I should investigate more.

File: rero_ils/modules/api.py (hunk 2 above)
Comment (blankoworld) [reply rero_ils/modules/api.py]: I change it to `validation_error`

File: tests/ui/test_api.py (hunk 4 above)
Comment (blankoworld) [reply tests/ui/test_api.py]: Its comes from **ci**rculation **po**licies. My bad!

**Pull request merged**